CenterForOpenScience/osf.io
# [Improvement] Remember the page where user originates during ORCiD account creation and linking process [#OSF-7076]

## Purpose

If a user first links or creates an OSF account through ORCiD, the confirmation link always redirects to dashboard with the welcome message. The context which page the user comes from is lost during the process. For user login from Preprints, ERPC and Prereg, this can be confusing.

This PR fix the above issue by:
1. Retrieve and store `service_url` into the "anonymous session";
2. Check the above `service_url` and generate confirmation link with
   `destination=dashboard|prereg|erpc|osf-preprints` as query parameter
3. Redirect to the destination page after successful confirmation.

Here are some example url and landing page (if a new user is created)
- Prereg

```
    http://localhost:5000/confirm/external/54y4e
    /1jQ5Ww7d0mgPe3WLBqAAUceruE4Xwk/?destination=prereg
```
- ERPC

```
    http://localhost:5000/confirm/external/vqsyz/
    MJWaDWpx3cQvT3twuRVDoRbHb8AhCa/?destination=erpc
```
- Preprints

```
    http://localhost:5000/confirm/external/cdj6b
    /zL2tz2DsWzF3jzjpm0ajpi4TDytexj/?destination=osf-preprints
```
- OSF

```
    http://localhost:5000/confirm/external/xddag
    /NbeiWOkZgsT1uTUbvrX5txGf2rfK56/?destination=dashboard
```
## Changes

### QA Notes

* Test Flow:
  1. Go to OSF campaign register page (e.g. `[OSF Domain]/register/?campaign=prereg`), click "Already have an account?", and get redirected to CAS login page .
  2. Click "Sign in with ORCiD", authenticate with a **NEW** ORCiD account, and get redirected to OSF ORCiD register email page.
  3. Enter an email to link or create an account and receive the confirmation email.
  4. Verify that the confirmation email contains `destination` parameter to campaign.
  5. Verify that after click the confirmation link, you are redirected to the campaign landing page.
  6. Verify there is no OSF welcome message pop-up on the campaign landing page (for both "link" and "create".
* What's not covered (and will be fixed in https://openscience.atlassian.net/browse/CAS-1)
  1. Branded preprints is not covered.
  2. CAS login page does not display dedicated campaign message.
  3. Clicking "Create Account" on CAS login page does not redirect you back to the correct OSF campaign register page.

## Side effects
- [x] Improve/Update service ticket related tests in `test_cas_authentication.py` .
- [x] Welcome message? (Discussed with product team)
      no OSF welcome message is displayed when register and land on campaign landing page
- [x] Redirection other than campaign? (Discussed with product team)
      registration from campaign lands on campaign landing page
      registration from other places lands on osf home page
## Note

The behavior only affects ORCiD login. Similar behavior on normal OSF login/registration is implemented on ticket https://openscience.atlassian.net/browse/CAS-1.
## Ticket

https://openscience.atlassian.net/browse/OSF-7076
